### PR TITLE
Cache SwaggerInterfaceParsers when Possible

### PR DESCRIPTION
**Pull request by alzimmermsft**
# Description

Adds a new method to `RestProxyUtils` which is able to cache instances of `SwaggerInterfaceParser` if the `Class` and `SerializerAdapter` match a previous instance. Generally, every time a new service client is created a new instance of `SwaggerInterfaceParser` is created and the `SerializerAdapter` is usually the default instance so this will enable the same instance to be re-used. For example, a quick performance test using a static `BlobContainerClient` to get `BlobClient`s using `getBlobClient` saw an improvement of 1300ns/op to 900ns/op, let alone reducing the number of object in memory.

Another change made in this PR was removing `SerializerAdapter` as a property from both `SwaggerInterfaceParser` and `SwaggerMethodParser` as this shouldn't be part of their contracts. The parser classes are meant to manage information about a Swagger generated class or method, not parts of their functionality when used. Now, `SerializerAdapter` is taken as a parameter in APIs that required it and should only be managed by `RestProxy` which controls runtime behavior.

Lastly, the `SwaggerMethodParser` cache was turned into an instance-level cache instead of a static cache now that `SwaggerInterfaceParser` is cached. This should reduce the size of the cache, potentially helping with allowing it to be passed around and queried more quickly, as well as less thread contention.

# All SDK Contribution checklist:
- [x] **The pull request does not introduce [breaking changes]**
- [x] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md).**

## [General Guidelines and Best Practices](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#developer-guide)
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#building-and-unit-testing)
- [x] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: sdk/core/azure-core/src/main/java/com/azure/core/implementation/http/rest/RestProxyUtils.java [hunk 1]
```diff
@@ -220,4 +222,7 @@ public static HttpPipeline createDefaultPipeline() {
             .build();
     }
 
+    public static SwaggerInterfaceParser getSwaggerInterfaceParser(Class<?> swaggerInterface) {
+        return INTERFACE_PARSERS.computeIfAbsent(swaggerInterface, SwaggerInterfaceParser::new);
+    }
```
Comment (kasobol-msft) [inline sdk/core/azure-core/src/main/java/com/azure/core/implementation/http/rest/RestProxyUtils.java:227]: I'd move this to `SwaggerInterfaceParser`. We can have two layers of caching there. I think it's better to keep them in one place.

File: sdk/core/azure-core/src/main/java/com/azure/core/implementation/http/rest/SwaggerInterfaceParser.java [hunk 2]
```diff
@@ -63,20 +40,20 @@ public SwaggerInterfaceParser(Class<?> swaggerInterface, SerializerAdapter seria
...
+     * Get the method parser that is associated with the provided swaggerMethod. The method parser can be used to get
+     * details about the Swagger REST API call.
      *
      * @param swaggerMethod the method to generate a parser for
      * @return the SwaggerMethodParser associated with the provided swaggerMethod
      */
     public SwaggerMethodParser getMethodParser(Method swaggerMethod) {
-        return METHOD_PARSERS.computeIfAbsent(swaggerMethod, sm ->
-             new SwaggerMethodParser(sm, getHost(), serializer));
+        return METHOD_PARSERS.computeIfAbsent(swaggerMethod, sm -> new SwaggerMethodParser(this, sm, getHost()));
```
Comment (kasobol-msft) [inline sdk/core/azure-core/src/main/java/com/azure/core/implementation/http/rest/SwaggerInterfaceParser.java:50]: as discussed offline, let's make it instance level cache.

File: sdk/core/azure-core/src/main/java/com/azure/core/implementation/http/rest/SwaggerMethodParser.java [hunk 3]
```diff
@@ -470,9 +478,11 @@ public UnexpectedExceptionInformation getUnexpectedException(int code) {
      * Get the object to be used as the value of the HTTP request.
      *
      * @param swaggerMethodArguments the method arguments to get the value object from
+     * @param serializer {@link SerializerAdapter} used to encode the request body if it's an
+     * {@code application/x-www-form-urlencoded} request.
      * @return the object that will be used as the body of the HTTP request
      */
-    public Object setBody(Object[] swaggerMethodArguments) {
+    public Object setBody(Object[] swaggerMethodArguments, SerializerAdapter serializer) {
```
Comment (srnagar) [inline sdk/core/azure-core/src/main/java/com/azure/core/implementation/http/rest/SwaggerMethodParser.java:485]: Do we need to pass the serializerAdapter for every method? If so, should we just make this a param in `getSwaggerInterfaceParser` like we had before in the ctor?

Comment (alzimmermsft) [reply sdk/core/azure-core/src/main/java/com/azure/core/implementation/http/rest/SwaggerMethodParser.java:485]: No, `SerializerAdapter` isn't needed for each method, nor does it exist in `SwaggerInterfaceParser` anymore

File: sdk/core/azure-core/src/main/java/com/azure/core/implementation/http/rest/SwaggerInterfaceParser.java [hunk 4]
```diff
@@ -5,53 +5,43 @@
...
+     * Create a SwaggerInterfaceParser object with the provided fully qualified interface name.
+     *
      * @param swaggerInterface The interface that will be parsed.
-     * @param serializer The serializer that will be used to serialize non-String header values and query values.
+     * @return The {@link SwaggerInterfaceParser} for the passed interface.
      */
-    public SwaggerInterfaceParser(Class<?> swaggerInterface, SerializerAdapter serializer) {
-        this(swaggerInterface, serializer, null);
-    }
+    public static SwaggerInterfaceParser getOrCreateInterfaceParser(Class<?> swaggerInterface) {
```
Comment (kasobol-msft) [inline sdk/core/azure-core/src/main/java/com/azure/core/implementation/http/rest/SwaggerInterfaceParser.java:31]: nit: I'd call it `getInstance`

**Review by kasobol-msft — APPROVED** [review]

**Pull request merged**